packages/twenty-front/src/modules/workflow/components/WorkflowEffect.tsx

```diff
@@ -0,0 +1,81 @@
+import { CoreObjectNameSingular } from '@/object-metadata/types/CoreObjectNameSingular';
+import { useCreateOneRecord } from '@/object-record/hooks/useCreateOneRecord';
+import { useWorkflowWithCurrentVersion } from '@/workflow/hooks/useWorkflowWithCurrentVersion';
+import { workflowDiagramState } from '@/workflow/states/workflowDiagramState';
+import { workflowIdState } from '@/workflow/states/workflowIdState';
+import { WorkflowVersion } from '@/workflow/types/Workflow';
+import { addCreateStepNodes } from '@/workflow/utils/addCreateStepNodes';
+import { getWorkflowVersionDiagram } from '@/workflow/utils/getWorkflowVersionDiagram';
+import { useEffect, useState } from 'react';
+import { useSetRecoilState } from 'recoil';
+import { isDefined } from 'twenty-ui';
+
+type WorkflowEffectProps = {
+  workflowId: string;
+};
+
+export const WorkflowEffect = ({ workflowId }: WorkflowEffectProps) => {
+  const workflowWithCurrentVersion = useWorkflowWithCurrentVersion(workflowId);
+
+  const setWorkflowId = useSetRecoilState(workflowIdState);
+  const setWorkflowDiagram = useSetRecoilState(workflowDiagramState);
+
+  const [isCreatingInitialVersionState, setIsCreatingInitialVersionState] =
+    useState<boolean>(false);
+
+  const { createOneRecord: createOneWorkflowVersion } =
+    useCreateOneRecord<WorkflowVersion>({
+      objectNameSingular: CoreObjectNameSingular.WorkflowVersion,
+    });
+
+  useEffect(() => {
+    setWorkflowId(workflowId);
+  }, [setWorkflowId, workflowId]);
+
+  useEffect(() => {
+    setIsCreatingInitialVersionState(false);
+  }, [workflowId]);
+
+  useEffect(() => {
+    if (!isDefined(workflowWithCurrentVersion)) {
+      setWorkflowDiagram(undefined);
+
+      return;
+    }
+
+    const currentVersion = workflowWithCurrentVersion.currentVersion;
+    if (!isDefined(currentVersion)) {
+      setWorkflowDiagram(undefined);
+
+      if (isCreatingInitialVersionState) {
+        return;
+      }
+
+      setIsCreatingInitialVersionState(true);
+
+      createOneWorkflowVersion({
+        workflowId,
+        status: 'DRAFT',
+        name: 'v1',
+      }).finally(() => {
+        setIsCreatingInitialVersionState(false);
+      });
+
+      return;
+    }
+
+    const lastWorkflowDiagram = getWorkflowVersionDiagram(currentVersion);
+    const workflowDiagramWithCreateStepNodes =
+      addCreateStepNodes(lastWorkflowDiagram);
+
+    setWorkflowDiagram(workflowDiagramWithCreateStepNodes);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I had to disable this ESLint rule because the reference of the `createOneWorkflowVersion` is unstable and changes at each render.

Many people use a function like this to circumvent this usual problem with React: https://github.com/scottrippey/react-use-event-hook.

Reply:
Think you can use a `useMemo` hook for `createOneWorkflowVersion` to avoid that

Reply:
Could you elaborate, please? What code do you have in mind?

Reply:
something like:
`
const createWorkflowVersion = useMemo(
if(isDefined(createOneWorkflowVersion)) {
return (data: {...}) => createOneWorkflowVersion(data)
}
, [createOneWorkflowVersion]
)
`
Then use `createWorkflowVersion` and add `createWorkflowVersion` in the useEffect dependencies, and remove the eslint comment
Not entirely sure it works but we can give it a try

Reply:
The issue is that the `createOneRecord` function returned by the `useCreateOneRecord` hook is unstable between renders; the hook returns a new function instance at each render. That makes it impossible to use the function in the dependency array of `useEffect` and `useMemo` hooks; otherwise, the hook might end up in an infinite loop.

See how the `createOneRecord` function is defined: https://github.com/twentyhq/twenty/blob/165df8959d18fb24c2c607c187e1189f856e4eb2/packages/twenty-front/src/modules/object-record/hooks/useCreateOneRecord.ts#L56.

To prevent bugs like this, the `createOneRecord` function should be memoized in the `createOneWorkflowVersion` hook with a `useCallback`. However, the `createOneRecord` function depends on many states that are likely to change often, which would force the `useCallback` to re-run and to return a new function instance.

For that kind of functions, I think it's usually better to return a stable reference, which can be achieved with a function like https://github.com/scottrippey/react-use-event-hook, or by relying on the React `useRef` hook.

For instance, the ApolloClient library for React uses references for hooks like `useMutation` to return stable references for function: https://github.com/apollographql/apollo-client/blob/1f0333e1644cb95b6862441ff957e60ec5f7a59e/src/react/hooks/useMutation.ts#L72

See the ref that is used to share data across callbacks without triggering re-renders:
https://github.com/apollographql/apollo-client/blob/1f0333e1644cb95b6862441ff957e60ec5f7a59e/src/react/hooks/useMutation.ts#L94

The `useCallback` callbacks have empty dependency arrays.

For the exec mutation function: 
https://github.com/apollographql/apollo-client/blob/1f0333e1644cb95b6862441ff957e60ec5f7a59e/src/react/hooks/useMutation.ts#L218

For the reset function:

https://github.com/apollographql/apollo-client/blob/1f0333e1644cb95b6862441ff957e60ec5f7a59e/src/react/hooks/useMutation.ts#L231

Reply:
Thanks for the exhaustive explanation. Indeed, there might be an infinite useMemo call, even if protected with the isDefined check it is not a proper solution to this issue. Ok to leave the eslint comment